MovieTime/MovieTime/Service/MTMovieDBService.swift

```diff
@@ -0,0 +1,87 @@
+//
+//  MTMovieDBService.swift
+//  MovieTime
+//
+//  Created by Enrique Melgarejo on 08/01/18.
+//  Copyright © 2018 Choynowski. All rights reserved.
+//
+
+import Foundation
+
+protocol MTMovieDBServiceProtocol: class {
+    init(service: MTServiceProtocol, helper: MTServiceHelperProtocol)
+
+    func requestConfiguration(completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestGenresList(completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestUpcomingMovies(forPage page: Int, completion: ((Data?, MTNetworkError?) -> Void)?)
+    func requestMovieDetail(forId movieId: Int, completion: ((Data?, MTNetworkError?) -> Void)?)
+}
+
+final class MTMovieDBService: MTMovieDBServiceProtocol {
+
+    private let urlServer = URL(string: "https://api.themoviedb.org/3")
+    private let currentService: MTServiceProtocol
+    private let serviceHelper: MTServiceHelperProtocol
+
+    private let apiKey = "1f54bd990f1cdfb230adb312546d765d"
+    private let language = "en-US"
+
+    init(service: MTServiceProtocol = MTService(), helper: MTServiceHelperProtocol = MTServiceHelper()) {
+        currentService = service
+        serviceHelper = helper
+    }
+
+    func requestConfiguration(completion: ((Data?, MTNetworkError?) -> Void)?) {
+
+        let endpoint = "/configuration"
+        let parameters: MTParameters = ["api_key": apiKey,
+                                        "language": language]
+        let url = urlServer?.appendingPathComponent(endpoint)
+
+        executeRequest(url: url, method: .get, parameters: parameters, completion: completion)
+    }
+
+    func requestGenresList(completion: ((Data?, MTNetworkError?) -> Void)?) {
+
+        let endpoint = "/genre/movie/list"
+        let parameters: MTParameters = ["api_key": apiKey,
+                                        "language": language]
+        let url = urlServer?.appendingPathComponent(endpoint)
+
+        executeRequest(url: url, method: .get, parameters: parameters, completion: completion)
```

Comment:
how would you refactor these `request*` methods to reduce the repeated code among them?

Reply:
The "api_key" and "language" parameters could be inside the `executeRequest (...)` method, but the rest of the information is unique to each request, such as: HttpMethod, parameters and headers.

Reply:
this line is common as well: `let url = urlServer?.appendingPathComponent(endpoint)`

IMO there is more redundancy than particularities, so refactoring these methods into smaller ones that call a more generic one would help you to have a more concise API and better code coverage with less tests